src/main/java/com/aws/greengrass/componentmanager/ComponentServiceHelper.java

```diff
@@ -65,8 +72,24 @@ ResolvedComponentVersion resolveComponentVersion(String componentName, Semver lo
                 .componentCandidates(Collections.singletonList(candidate)).build();
 
         ResolveComponentCandidatesResponse result;
-        try {
-            result = clientFactory.getGreengrassV2DataClient().resolveComponentCandidates(request);
+
+        RetryUtils.RetryConfig clientExceptionRetryConfig =
+                RetryUtils.RetryConfig.builder().initialRetryInterval(Duration.ofSeconds(30))
+                        .maxRetryInterval(Duration.ofSeconds(30)).maxAttempt(CLIENT_RETRY_COUNT)
+                        .retryableExceptions(Arrays.asList(DeviceConfigurationException.class)).build();
+
+        try (GreengrassV2DataClient greengrasV2DataClient = RetryUtils.runWithRetry(clientExceptionRetryConfig, () -> {
+            GreengrassV2DataClient client = clientFactory.getGreengrassV2DataClient();
+            if (client == null) {
+                String errorMessage = clientFactory.getConfigValidationError().isPresent()
+                        ? clientFactory.getConfigValidationError().get() : "Could not get GreengrassV2DataClient";
```

Comment:
nit: you could combine this as `clientFactory.getConfigValidationError().orElse("Could not get GreengrassV2DataClient")`

Reply:
That's cleaner. Updated.